src/hooks/useAddressInput.ts

```diff
@@ -0,0 +1,33 @@
+import React from "react";
+import { isValidAddress } from "../utils/address";
+import { useConnection } from "./useConnection";
+
+export function useAddressInput() {
+  const { account } = useConnection();
+
+  const [address, setAddress] = React.useState(account ? account : "");
+  const prevAddress = React.useRef(address);
+
+  const updateAddress = React.useCallback(
+    (newAddress: string) => {
+      prevAddress.current = address;
+      setAddress(newAddress);
+    },
+    [address]
```

Comment:
this function is a bit suspect, i cant quite put my finger on it.  the useCallback is almost useless since when you are typing, its re-instancing the function for every key press since you are listeining for the address change. Its also kind of recursively doing this, since inside the callback is triggering the change to the callbacks `address`.   

Why does it need to listen to address?

Reply:
I agree the `useCallback` doesn't save much in terms of re-allocations, so if you think its confusing, I will remove it. 

Regarding the function itself, a new function needs to be allocated each time `address` changes, as it as closure over it and would reference stale values if this was not the case.
For example, if the function was only allocated once (with deps of `[]` to `useCallback`) then it would always reference the first value of `address` instead of the latest one. 
There is no recursiveness involved as when `setAddress` is called in `Hero` that triggers `Hero` to re-render, `useAddressInput` gets run again, and `updateAddress` get re-allocated, with closure over the latest `address` value. 

If no `useCallback` was present, it would be re-allocated at every re-render regardless of wether `address` changed, achieving the same effect as above, just with a few more renders.

Reply:
hmm yea i mean i guess the use callback is better than not having it.  what about the prevAddress ref, i can see its there to switch between displaying the logged in account vs a user entered account, but i cant see it doing much more than just signifying when the user has entered data, ie a "dirty bit".  How is what you have different from this simplification:

```js
const [address, setAddress] = React.useState("");
const updateAddress = setAddress
React.useEffect(() => {
    if (!account) return;
    if(address !="") return;
     setAddress(account);
  }, [account]);

```